### PR TITLE
Python

### PR DESCRIPTION
**Pull request by markus2330**
## Basics

Do not describe the purpose here but:

- [ ] Short descriptions should be in the release notes (added as entry in
      doc/news/_preparation_next_release.md which contains `*(my name)*`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins)

## Review

Remove the line below and add the "work in progress" label if you do
not want the PR to be reviewed:

@markus2330 please review my pull request

### CODE REVIEW

**Comment by markus2330** [comment]
@tom-wa this should make the python plugin stable. I have not tested it, though.

**Comment by markus2330** [comment]
@ingwinlu Do you know how we can make the python unit tests locate their respective "kdb module" (both version 2 and 3). I still do not understand why we do not install Elektra to /usr/local, wouldn't that be much easier? Then such things would work out of the box.

**Comment by markus2330** [comment]
@e1528532 did I use pluginprocess correctly?

I do not quite understand why ELEKTRA_PLUGINPROCESS_OPEN gets implicitely called but not ELEKTRA_PLUGINPROCESS_CLOSE? And in line 384 of src/libs/pluginprocess/pluginprocess.c, shouldn't that be ELEKTRA_PLUGINPROCESS_CLOSE instead of ELEKTRA_PLUGIN_CLOSE?

**Comment by markus2330** [comment]
jenkins build all please

**Comment by markus2330** [comment]
@sanssecours Do you have any idea why "swig3.0 -c++ -python -external-runtime runtime.h" does not produce runtime.h under Mac OS X? Or is there some other reason why https://travis-ci.org/ElektraInitiative/libelektra/jobs/409880622 fails?

@ingwinlu Seems like "Pull docker images" hangs?

**Comment by ingwinlu** [comment]
nothing hangs. 36 builds in queue as we have taken down v2 for benchmarking one of our plugins.

**Comment by markus2330** [comment]
> nothing hangs

?

> 36 builds in queue

How does the build server come up with this high number, these are much more jobs than PRs which had activity today.

Would be great if this PR can be merged today, I need Debian packages of it and unfortunately there is no easy way to create Debian packages without merging to master.

**Comment by ingwinlu** [comment]
> How does the build server come up with this high number, these are much more jobs than PRs which had activity today.

because our builds run in parallel. each parallel stage adds 1 to the number.

>  I need Debian packages of it 

pretty sure master is broken since the yaml merge.

> unfortunately there is no easy way to create Debian packages without merging to master.

that was a compromise to not have even more builds that the system can't handle. It can be rewritten to build all the time with the obvious downside of slowing down the overworked system even more.

**Comment by markus2330** [comment]
> that was a compromise to not have even more builds that the system can't handle.

I thought the multiple repos are the problem. (PRs should not overwrite the packages of master.)

> It can be rewritten to build all the time with the obvious downside of slowing down the overworked system even more.

This can be fixed with more hardware. When v2 is running we usually do not have any problems anyway.

**Comment by sanssecours** [comment]
> Do you have any idea why "swig3.0 -c++ -python -external-runtime runtime.h" does not produce runtime.h under Mac OS X?

Nope, sorry.

> Or is there some other reason why https://travis-ci.org/ElektraInitiative/libelektra/jobs/409880622 fails?

Not as far as I can tell.

**Comment by markus2330** [comment]
>   Nope, sorry.

What happens when the above command is executed? Maybe we use a wrong version of swig? With "3.0.10" on Linux it works without problems.

The alternative would be to disable the python plugins for Mac OS X...

**Comment by sanssecours** [comment]
> What happens when the above command is executed? Maybe we use a wrong version of swig?

The Homebrew formula of SWIG does not include a binary called `swig3.0`:

```sh
find (brew --cellar swig) -regex '.*bin.*'
#> /usr/local/Cellar/swig/3.0.12/bin
#> /usr/local/Cellar/swig/3.0.12/bin/swig
#> /usr/local/Cellar/swig/3.0.12/bin/ccache-swig
```

.

> The alternative would be to disable the python plugins for Mac OS X...

An easier alternative would be to just remove the [Python configuration](https://master.libelektra.org/.travis.yml#L247-L255) from `.travis.yml`.

**Comment by markus2330** [comment]
> The Homebrew formula of SWIG does not include a binary called swig3.0

Might also be called differently, SWIG_EXECUTABLE of CMakeCache.txt should tell you which one is used.

>  An easier alternative would be to just remove the Python configuration from .travis.yml.

Even if we know it does not compile? Its status reads "maintained unittest configurable global memleak" so people would not expect it to not work.

**Comment by markus2330** [comment]
Ok, then let us disable the python binding for the GCC build in Mac OSX. Thank you for your investigation!

File: doc/news/_preparation_next_release.md [hunk 1]
```diff
@@ -164,7 +164,7 @@ Thanks to Daniel Bugl.
 ### Interpreter Plugins
 
 - The plugins ruby, python and jni can now also be mounted as global plugin.
-- Fix crash in python plugin. *(Markus Raab)*
+- Fix crash in python plugin by using process plugin. *(Markus Raab)*
```
[the remark this answers is not in the document]
Comment (markus2330) [reply doc/news/_preparation_next_release.md]: Thank you! My initial plan was to use the process plugin but this changed now.

Can you maybe create a new PR fixing the pluginprocess #2142? If it is useful for you can cherry-pick my commits in this PR.

File: src/plugins/python/python.cpp [hunk 2]
```diff
@@ -325,6 +370,15 @@ int PYTHON_PLUGIN_FUNCTION (Close) (ckdb::Plugin * handle, ckdb::Key * errorKey)
 	moduleData * data = static_cast<moduleData *> (elektraPluginGetData (handle));
 	if (data == nullptr) return 0;
 
+	ElektraPluginProcess * pp = data->pp;
+	if (elektraPluginProcessIsParent (pp))
+	{
+		int result = elektraPluginProcessSend (pp, ELEKTRA_PLUGINPROCESS_CLOSE, NULL, errorKey);
+		if (elektraPluginProcessClose (pp, errorKey)) elektraPluginSetData (handle, NULL);
```
[the remark this answers is not in the document]
Comment (markus2330) [reply src/plugins/python/python.cpp]: It was copied from the example. Shouldn't elektraPluginProcessClose be called in both parent & child?

**Comment by markus2330** [comment]
> please feel free to correct the issue

It would be great if you can do the necessary fixes processplugin.

> i haven't tried it out yet though on my own machine, i assume it should work?

It improves one specific problem but I am afraid there are other problems with the current protocol. The library also needs more unit tests.

File: src/plugins/python/python.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (e1528532) [reply src/plugins/python/python.cpp]: well the child process gets killed anyway so it shouldn't make much of a difference

[the remark this answers is not in the document]
Comment (e1528532) [reply src/plugins/python/python.cpp]: ok the child does call cleanupPluginData at the end of elektraPluginProcessStart so it should be fine.

**Comment by markus2330** [comment]
I think changing the header is okay because I do not expect the library is used externally at the moment. (Just mention it in the release notes.) If you think there is potential to improve how the library can be used (less code in the plugins) please go for it. If the change is only marginal then better keep it like it is.

You can also directly modify the python branch. I would appreciate it if you can fix the python plugin.

**Comment by e1528532** [comment]
modified the branch. Pluginprocess seems to work fine now, however if you debug it you notice that the python interpreter startup fails / gets stuck. Seems to happen somewhere there:

```
		/* acquire GIL */
		Python_LockSwap pylock (nullptr);

		/* create a new sub interpreter */
		data->tstate = Py_NewInterpreter ();
		if (data->tstate == nullptr)
		{
			ELEKTRA_SET_ERROR (111, errorKey, "Unable to create sub interpreter");
			goto error;
		}
		PyThreadState_Swap (data->tstate);
```
i have to say though that i don't really want to mess around with python at this point as i can't predict how much time it will take and i have the august deadline filling my mind.. However it should be fairly easy to debug pluginprocess when you enable the logs to check whether python or pluginprocess gets stuck.

**Comment by markus2330** [comment]
For me the tests also get stuck and they did not get stuck before. So it is related to your change.

You do not have to mess with python internals but it would be great if the processplugin is used correctly within the python plugin.

**Comment by e1528532** [comment]
hmm. ok, i will try to find something out then though for me it gets stuck inside the python initialization now.

**Comment by markus2330** [comment]
Maybe the python initialization is called twice? E.g. first in parent and then again in child? Python is really unreliable in such situations.

**Comment by e1528532** [comment]
Locally it seems to work now with pluginprocess:

```
The following tests passed:
	testmod_python
```

though elektra has to be installed otherwise it won't find the kdb module:

```
		/* extend sys path for kdb module */
		ELEKTRA_LOG ("it is located at %s", ELEKTRA_PYTHON_SITE_PACKAGES);
```

printed `/usr/local/python3.6/site-packages` to me. Furthermore i removed this "keep python open" logic as to my understanding this was added because pluginprocess was missing? Now every instantiation of a plugin should happen in its own process so it should be fine to restart the interpreter all the time. I suspect `data->shutdown` is a leftover of that however i am not sure about it.

In any case if the pluginprocess integration indeed works now, as it seems locally, i'd pass this PR back to you @markus2330 as the pluginprocess improvements already ate a lot of time and end of august is near if thats ok.

so lets see what the build server says. I hope the pluginprocess asan issues are fixed now,

**Comment by markus2330** [comment]
Thank you for the update!

The last commit seems to lead to segfaults because then kdbGet might be called on non-initialized pluginprocess. How is this done for the haskell plugin?

**Comment by markus2330** [comment]
kdbOpen is called, but with /module.

**Comment by e1528532** [comment]
I think the pluginprocess-related issues should now really be gone.

The current error is the one that i've mentioned, that it expects the kdb python module to be installed to the system? `reason: : Unable to import kdb module`. Not sure if its related to pluginprocess as testmod_python and such work and there is no segfault due to pluginprocess either.

**Comment by markus2330** [comment]
Let us merge #2149 first.

**Comment by markus2330** [comment]
@e1528532 Can you rebase this PR (in particular remove your commits and instead rebase to master)?

And is your rejected review still up-to-date?

**Comment by e1528532** [comment]
yes i've worked in my review remarks or they are no longer required after the pluginprocess maintenance works. Will rebase and then I'd give the PR back to you because i really need to focus on the thesis.

**Comment by e1528532** [comment]
@markus2330 there you go

**Comment by markus2330** [comment]
Finally it is done :+1: 

If someone is interested why there are two travis jobs, one is because of ["Build pushed branches"](https://docs.travis-ci.com/user/web-ui/#build-pushed-pull-requests) and the other because of the pull request ([branches merged into master](https://docs.travis-ci.com/user/pull-requests/#double-builds-on-pull-requests)).

**Pull request merged**